File: 383.Ransom-Note.go
// https://leetcode.com/problems/ransom-note/
//
// algorithms
// Medium (48.96%)
// Total Accepted:    99,221
// Total Submissions: 202,642
// beats 62.16% of golang submissions

package leetcode

func canConstruct(ransomNote string, magazine string) bool {
	hashMapNote, hashMapMagazine := make(map[rune]int), make(map[rune]int)

	for _, ch := range ransomNote {
		if v, ok := hashMapNote[ch]; ok {
			hashMapNote[ch] = v + 1
		} else {
			hashMapNote[ch] = 1
		}
	}

	for _, ch := range magazine {
		if v, ok := hashMapMagazine[ch]; ok {
			hashMapMagazine[ch] = v + 1
		} else {
			hashMapMagazine[ch] = 1
		}
	}

	for k, v := range hashMapNote {
		if v1, ok := hashMapMagazine[k]; !ok || v1 < v {
			return false
		}
	}

	return true
}

File: 383.Ransom-Note.py
# https://leetcode.com/problems/ransom-note/
#
# algorithms
# Easy (50.76%)
# Total Accepted:    124,538
# Total Submissions: 245,370


class Solution(object):
    def canConstruct(self, ransomNote, magazine):
        """
        :type ransomNote: str
        :type magazine: str
        :rtype: bool
        """
        chs = [0] * 26
        for ch in magazine:
            idx = ord(ch) - 97
            chs[idx] += 1

        for ch in ransomNote:
            idx = ord(ch) - 97
            chs[idx] -= 1
            if chs[idx] < 0:
                return False

        return True
